src/main/java/net/minecraftforge/client/extensions/IForgeAtlasTexture.java

```diff
@@ -0,0 +1,197 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.client.extensions;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.texture.AtlasTexture;
+import net.minecraft.client.renderer.texture.TextureAtlasSprite;
+import net.minecraft.resources.IResourceManager;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.Util;
+import net.minecraftforge.client.event.TextureStitchEvent;
+import net.minecraftforge.client.model.ModelDynBucket;
+import net.minecraftforge.client.model.ModelLoader;
+import net.minecraftforge.fml.ModLoader;
+import net.minecraftforge.fml.client.ClientHooks;
+import net.minecraftforge.fml.loading.toposort.CyclePresentException;
+import net.minecraftforge.fml.loading.toposort.TopologicalSort;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public interface IForgeAtlasTexture
+{
+    Logger LOGGER = LogManager.getLogger();
+
+    default AtlasTexture getAtlasTexture()
+    {
+        return (AtlasTexture)this;
+    }
+
+    @FunctionalInterface
+    interface SpriteProvider
+    {
+        TextureAtlasSprite create(IResourceManager manager) throws Exception;
+    }
+
+    default List<TextureAtlasSprite> resolveSpriteDependencies(Collection<TextureAtlasSprite> sprites)
+    {
+        Map<ResourceLocation, TextureAtlasSprite> idToSprite = Maps.newHashMap();
+        MutableGraph<TextureAtlasSprite> graph = GraphBuilder.directed().allowsSelfLoops(false).build();
+        sprites.forEach(node -> {
+            graph.addNode(node);
+            idToSprite.put(node.getName(), node);
+        });
+        sprites.forEach(s -> s.getDependencies().forEach(dep -> {
+            final TextureAtlasSprite dependency = idToSprite.get(dep);
+            if (dependency == null)
+            {
+                ClientHooks.trackBrokenTexture(dep, "Missing dependency requested from sprite " + s.getName());
+            }
+            else
+            {
+                graph.putEdge(dependency, s);
+            }
+        }));
+
+        try
+        {
+            return TopologicalSort.topologicalSort(graph, null);
+        }
+        catch (CyclePresentException e)
+        {
+            Set<Set<TextureAtlasSprite>> cycles = e.getCycles();
+            cycles.stream().flatMap(Set::stream).distinct().forEach(sprite -> {
```

Comment:
The `distinct()` call should be extraneous as sets don't contain duplicate elements by their contract, though this does not harm performance in any way as the stream from a set probably already has the distinct flag.

Reply:
`distinct` is on result on `flatMap`. While I know that those sets are disjoint, since this actually returns strongly connected sets, method is just called `getCycles`, so I can't assume anything about that. Also, this is only called once during exception handling, so not really hot spot.